keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,823 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Adds a migration path to use server-side-apply from usage of client-side-apply.
+
+## Motivatio
+
+The current documentation for Kubernetes Server-Side-Apply has instructions for
+migrating from usage of client-side-apply to server-side-apply:
+
+QUOTE
+
+
+Unfortunately it is not so simple. By following these instructions you can be
+left with the following situation:
+
+```
+$ csa an object
+```
+
+```
+$ ssa same object
+```
+
+```
+$ ssa object but delete field
+```
+
+```
+$ show field is still present, despite last command, due to csa manager retaining ownership
+```
+
+This KEP is intended to bridge the gap, and provide a one-time operation
+
+### Goals
+
+1. Provide irreversable operation to kubectl users to convert management of an object's fields
+from client-side-apply to server-side-apply
+
+
+### Non-Goals
+
+1. Provide a server-side fix to this problem
+
+## Proposal
+
+Add a new CLI flag `--csa-manager`.
+
+When --server-side is used, fields owned by ---csa-manager will be converted to
+ownership of the ssa manager.
+
+If --csa-manager is not specified, then it is implied to be the same value as
+`--fieldmanager`
+
+
+```
+# Apply object with client-side-apply
+kubectl apply -f pod.yaml --fieldmanager mycsamanager
+```
+
+```
+# Apply object again with server-side-apply
+kubectl apply -f pod.yaml --fieldmanager=myssamanager --csa-manager=mycsamanager --server-side
+```
+
+With these flags, kubectl will migrate all fields owned by `mycsamanager` for client-side-apply
+to be now be owned by `myssamanager` for use with server-side-apply. This process
+is irreversible.
+
+### Inferred Defaults
+
+```
+# Apply object with client-side-apply
+kubectl apply -f pod.yaml
+```
+
+```
+# Apply object again with server-side-apply
+# Behind the scenes --csa-manager is inferred to be "kubectl-client-side-apply", 
+# and its fields are migrated to server-side-apply
+kubectl apply -f pod.yaml --server-side
```

Comment:
is this doing the conversion even if I didn't specify the csa-manager? if that's the case, and that we need an extra request to do the conversion, are we going to make 2 calls every single time we apply?

Reply:
IIRC kubectl is unconditionally doing RMW when you apply even with server-side. Assuming my memory is correct, and assuming we want to keep that behavior this would not be any additional requests.

You are right however, that the initial read for every apply is unnecessary for those not doing upgrades. I was thinking it would be optimal to keep the number of steps the users have to learn at 0, but am not opposed to requiring `--csa-manager` rather than having implicit defaults.

Reply:
You can't do a RMW with server-side apply actually (and should absolutely not). Today we get the object, but we don't use it at all for server-side apply. So it's still mean that the process is twofold, first migrate then apply, which means it's not obvious that we should do this automatically for every occurrences of apply